File: contrib/Examples/python/examples.py
import sys
import numpy as np
import arlpy as ap
from unetpy import *

################### Open connection to modem #################################
# Get the IP address of the modem from argument passed
if len(sys.argv) < 2:
    print("Please provide the IP address of the modem as argument!")
    sys.exit(0)

ip_address = sys.argv[1]
# Open a connection to modem
modem = UnetGateway(ip_address, 1100)

################### Transmission of CONTROL and DATA packet ###################

# Subscribe to receive physical agent notification
phy = modem.agentForService(Services.PHYSICAL)

# Transmit a CONTROL packet
phy << TxFrameReq(type=Physical.CONTROL, data=np.arange(4))
# Receive the transmit notification
txntf1 = modem.receive(TxFrameNtf, 5000)
if txntf1 is not None:
    # Extract the transmission start time
    txtime = txntf1.txTime
    print('The transmission started at ' + str(txtime))
else:
    print('Transmission not successfull, try again!')

# Transmit a DATA packet
phy << TxFrameReq(type=Physical.DATA, data=np.arange(10))
# Receive the transmit notification
txntf2 = modem.receive(TxFrameNtf, 5000)
if txntf2 is not None:
    # Extract the transmission start time
    txtime = txntf2.txTime
    print('The transmission started at ' + str(txtime))
else:
    print('Transmission not successfull, try again!')

################### Transmission of baseband signal ##########################

# Subscribe to receive baseband agent notification
bb = modem.agentForService(Services.BASEBAND)

# Generate a random signal modulated with BPSK
csig = ap.comms.modulate(ap.comms.random_data(10000), ap.comms.psk())
# Convert complex signal (csig) with real and imag to an array
# with alternate real and imag values
real_csig = csig.real
imag_csig = csig.imag
signal = np.empty((real_csig.size + imag_csig.size,), dtype=real_csig.dtype)
signal[0::2] = real_csig
signal[1::2] = imag_csig
# Transmit the baseband signal
bb << TxBasebandSignalReq(preamble=3, signal=signal)
# Receive the transmit notification
txntf3 = modem.receive(TxFrameNtf, 5000)
if txntf3 is not None:
    # Extract the transmission start time
    txtime = txntf3.txTime
    print('The transmission started at ' + str(txtime))
else:
    print('Transmission not successfull, try again!')

################### Recording a baseband signal #################################

# Record a baseband signal
bb << RecordBasebandSignalReq(recLen=24000)
# Receive the notification when the signal is recorded
rxntf = modem.receive(RxBasebandSignalNtf, 5000)
if rxntf is not None:
    # Extract the recorded signal
    rec_signal = rxntf.signal
    print('Recorded signal successfully!')
    # The recorded signal saved in `rec_signal` variable
    # can be processed as required by the user.
else:
    print('Recording not successfull, try again!')

################### Close connection to modem ################################

modem.close()

File: contrib/Examples/c/examples.c
///////////////////////////////////////////////////////////////////////////////
// In terminal window (an example):
//
// > make test IP=localhost
//
// Set the actual IP address address of the modem when using modem.
//
// Find relevant documentation on UnetStack and APIs at the following link:
// https://www.unetstack.net/docs.html
////////////////////////////////////////////////////////////////////////////////

// TODO:
// change ip_address to hostname
// Add readme
// check recording again

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <math.h>
#include "include/unet.h"


static void txcb(const char* id, modem_packet_t type, long time) {
  printf("TxFrameNtf type:%d time:%ld id:%s\n", type, time, id);
}

static int error(const char* msg) {
  printf("\n*** ERROR: %s\n\n", msg);
  return -1;
}

static int tx(modem_t modem, int to, void* data, int nbytes, modem_packet_t type, char* id) {
  int x = modem_tx_data(modem, to, data, nbytes, type, id);
  if (x != 0) return -1;
  return 0;
}

static int tx_signal(modem_t modem, float* signal, int nsamples, float fc, char* id) {
  int x = modem_tx_signal(modem, signal, nsamples, fc, id);
  if (x != 0) return -1;
  return 0;
}

int main(int argc, char* argv[]) {
  char id[FRAME_ID_LEN];
  int data[7] = {1,2,3,4,5,6,7};
  int addressofDestination = (int)argv[2];
  float signal[20000] = {[0 ... 19999] = 1};
  float rec_signal[24000];

  /******************** Open connection to modem *********************************/

  // Open a connection to modem
  modem_t modem = modem_open_eth(argv[1], 1100);
  if (modem == NULL) return error("Couldn't open modem");
  modem_set_tx_callback(modem, txcb);

  /******************** Transmission of CONTROL and DATA packet ******************/

  // Transmit a CONTROL frame
  tx(modem, addressofDestination, data, 7, 1, id);
  sleep(3);

  // Transmit a DATA frame
  tx(modem, addressofDestination, data, 7, 2, id);
  sleep(3);

  /******************** Transmission of baseband signal ****************************/

  // Transmit a baseband signal
  tx_signal(modem, signal, 10000, 12000, id);
  sleep(3);

  /******************** Record of baseband signal **********************************/

  // Record a baseband signal
  if (modem_record(modem, rec_signal, 12000) < 0) {
    printf("Recording not successfull, try again!");
  } else {
    printf("Recorded signal successfully!");
    // The recorded signal saved in `rec_signal` variable
    // and can be processed as required by the user.
  }

  /******************** Close connection to modem ****************************/

  modem_close(modem);

  return 0;
}




